NDTensors/src/lib/BlockSparseArrays/src/blocksparsearray/blocksparsearray.jl

```diff
@@ -118,3 +118,8 @@ blockstype(::Type{<:BlockSparseArray{<:Any,<:Any,<:Any,B}}) where {B} = B
 ##   # TODO: Preserve GPU data!
 ##   return BlockSparseArray{elt}(undef, axes)
 ## end
+
+function Base.adjoint(A::BlockSparseMatrix)
+  T = Base.promote_op(adjoint, eltype(A))
```

Comment:
I'm confused by this line, why would `adjoint` change the element type of the matrix? Was this meant to by something like `Base.promote_op(adjoint, blocktype(A))`? If so, that doesn't seem necessary anyway since the block type will be inferred by the blocks that get input into the next constructor call.

Reply:
It is really intended like this, and is an implementation of
https://github.com/JuliaLang/julia/blob/5848445ea6ab572253768812d999e132d4f85638/stdlib/LinearAlgebra/src/adjtrans.jl#L64

I guess the idea is that you have absolutely no control over what element types are valid, and if you want to ensure that `getindex` returns something of type `T`, this has to be checked. Ofcourse, for most types this is just an identity operation, but a counterexample (slightly contrived) is to have a blockarray consisting of an array of arrays of arrays, in which case the `eltype` was `Array` before, and `Adjoint` after.

Reply:
I see, that makes sense, thanks for clarifying.

Also, I think this is a nice definition, `Adjoint` wrapping `BlockSparseMatrix` has been annoying to deal with (https://github.com/ITensor/ITensors.jl/issues/1336), this should make that simpler. We may still want to support `Adjoint` wrapping `BlockSparseMatrix` but at least `adjoint` will create an object that is easier to deal with.

Reply:
yeah, I think it is a nice in-between in having the "lazy adjoint" behavior, while not having too many layers of wrappers.

Reply:
Could you move this into a separate PR that this PR is based on? That should help out @ogauthe.